state/protocol/protocol_state/state/protocol_state.go

```diff
@@ -103,34 +95,209 @@ func NewMutableProtocolState(
+//
+// EvolveState is compatible with speculative processing: it evolves an *in-memory copy* of the parent state
+// and collects *deferred database updates* for persisting the resulting Protocol State, including all of its
+// dependencies and respective indices. Though, the resulting batch of deferred database updates still depends
+// on the candidate block's ID, which is still unknown at the time of block construction. Executing the deferred
+// database updates is the caller's responsibility.
+//
+// SAFETY REQUIREMENTS:
+//  1. The seals must be a protocol-compliant extension of the parent block. Intuitively, we require that the
+//     seals follow the ancestry of this fork without gaps. The Consensus Participant's Compliance Layer enforces
+//     the necessary constrains. Analogously, the block building logic should always produce protocol-compliant
+//     seals.
+//     The seals guarantee correctness of the sealed execution result, including the contained service events.
+//     This is actively checked by the verification node, whose aggregated approvals in the form of a seal attest
+//     to the correctness of the sealed execution result (specifically the Service Events contained in the result
+//     and their order).
+//  2. For Consensus Participants that are replicas, the calling code must check that the returned `stateID` matches
+//     the commitment in the block proposal! If they don't match, the proposal is byzantine and should be slashed.
+//
+// Error returns:
+// [TLDR] All error returns indicate potential state corruption and should therefore be treated as fatal.
+//   - Per convention, the input seals from the block payload have already been confirmed to be protocol compliant.
+//     Hence, the service events in the sealed execution results represent the honest execution path.
+//     Therefore, the sealed service events should encode a valid evolution of the protocol state -- provided
+//     the system smart contracts are correct.
+//   - As we can rule out byzantine attacks as the source of failures, the only remaining sources of problems
+//     can be (a) bugs in the system smart contracts or (b) bugs in the node implementation. A service event
+//     not representing a valid state transition despite all consistency checks passing is interpreted as
+//     case (a) and _should be_ handled internally by the respective state machine. Otherwise, any bug or
+//     unforeseen edge cases in the system smart contracts would result in consensus halt, due to errors while
+//     evolving the protocol state.
+//   - A consistency or sanity check failing within the StateMutator is likely the symptom of an internal bug
+//     in the node software or state corruption, i.e. case (b). This is the only scenario where the error return
+//     of this function is not nil. If such an exception is returned, continuing is not an option.
+func (s *MutableProtocolState) EvolveState(
```

Comment:
`EvolveState` and secondary methods below replace the `stateMutator`. Conceptually, the logic has not changed (just arranged slightly differently)